File: android/src/main/kotlin/com/pycampers/flutter_cognito_plugin/serializer.kt
package com.pycampers.flutter_cognito_plugin

import com.amazonaws.mobile.client.UserStateDetails
import com.amazonaws.mobile.client.results.ForgotPasswordResult
import com.amazonaws.mobile.client.results.SignInResult
import com.amazonaws.mobile.client.results.SignUpResult
import com.amazonaws.mobile.client.results.UserCodeDeliveryDetails

fun dumpUserCodeDeliveryDetails(u: UserCodeDeliveryDetails?): List<String> {
    return if (u != null) {
        listOf(u.attributeName, u.destination, u.deliveryMedium)
    } else {
        listOf()
    }
}

fun dumpSignUpResult(signUpResult: SignUpResult): List<*> {
    return listOf(
            signUpResult.confirmationState
    ) + dumpUserCodeDeliveryDetails(
            signUpResult.userCodeDeliveryDetails
    )
}

fun dumpSignInResult(signInResult: SignInResult): List<*> {
    return listOf(
            signInResult.signInState.ordinal,
            signInResult.parameters
    ) + dumpUserCodeDeliveryDetails(
            signInResult.codeDetails
    )
}

fun dumpForgotPasswordResult(forgotPasswordResult: ForgotPasswordResult): List<*> {
    return listOf(
            forgotPasswordResult.state.ordinal
    ) + dumpUserCodeDeliveryDetails(
            forgotPasswordResult.parameters
    )
}

fun dumpUserState(userStateDetails: UserStateDetails): Int {
    return userStateDetails.userState.ordinal
}

File: ios/Classes/serializer.swift
import AWSMobileClient
import Flutter

func dumpUserCodeDeliveryDetails(_ userCodeDeliveryDetails: UserCodeDeliveryDetails?) -> [String] {
    if userCodeDeliveryDetails == nil {
        return []
    }

    let u = userCodeDeliveryDetails!
    return [
        u.attributeName!,
        u.destination!,
        String(describing: u.deliveryMedium)
    ]
}

func dumpSignUpResult(_ signUpResult: SignUpResult) -> [Any] {
    return [
        signUpResult.signUpConfirmationState == SignUpConfirmationState.confirmed
    ] + dumpUserCodeDeliveryDetails(
        signUpResult.codeDeliveryDetails
    )
}

let _androidSignInStateEnum = [
    SignInState.smsMFA,
    SignInState.passwordVerifier,
    SignInState.customChallenge,
    SignInState.deviceSRPAuth,
    SignInState.devicePasswordVerifier,
    SignInState.adminNoSRPAuth,
    SignInState.newPasswordRequired,
    SignInState.signedIn,
    SignInState.unknown
]

func dumpSignInResult(_ signInResult: SignInResult) -> [Any] {
    return [
        _androidSignInStateEnum.firstIndex(of: signInResult.signInState)!,
        signInResult.parameters
    ] + dumpUserCodeDeliveryDetails(
        signInResult.codeDetails
    )
}

let _androidForgotPasswordStateEnum = [
    ForgotPasswordState.confirmationCodeSent,
    ForgotPasswordState.done
]

func dumpForgotPasswordResult(_ forgotPasswordResult: ForgotPasswordResult) -> [Any] {
    return [
        _androidForgotPasswordStateEnum.firstIndex(of: forgotPasswordResult.forgotPasswordState)!
    ] + dumpUserCodeDeliveryDetails(
        forgotPasswordResult.codeDeliveryDetails
    )
}

let _androidUserStateEnum = [
    UserState.signedIn,
    UserState.guest,
    UserState.signedOutFederatedTokensInvalid,
    UserState.signedOutUserPoolsTokenInvalid,
    UserState.signedOut,
    UserState.unknown
]

func dumpUserState(_ userState: UserState) -> Int {
    return _androidUserStateEnum.firstIndex(of: userState)!
}

func dumpTokens(_ tokens: Tokens) -> [String?] {
    return [tokens.accessToken?.tokenString, tokens.idToken?.tokenString, tokens.refreshToken?.tokenString]
}

extension Date {
    var millisecondsSince1970: Int64 {
        return Int64((self.timeIntervalSince1970 * 1000.0).rounded())
    }
}

func dumpCredentials(_ credentials: AWSCredentials) -> [Any?] {
    return [
        credentials.accessKey,
        credentials.secretKey,
        credentials.sessionKey,
        credentials.expiration?.millisecondsSince1970
    ]
}
